File: hard/032-longest-valid-parentheses/solution.java
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * 32. Longest Valid Parentheses
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public int longestValidParentheses(String s) {
        int best = 0;
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(-1); // sentinel boundary before any valid run

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                stack.pop();
                if (stack.isEmpty()) {
                    stack.push(i); // unmatched ')', new boundary
                } else {
                    best = Math.max(best, i - stack.peek());
                }
            }
        }

        return best;
    }
}

File: hard/032-longest-valid-parentheses/solution.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        """
        32. Longest Valid Parentheses
        Time: O(n)
        Space: O(n)
        """
        best = 0
        stack = [-1]  # sentinel boundary before any valid run

        for i, ch in enumerate(s):
            if ch == "(":
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)  # unmatched ')', new boundary
                else:
                    best = max(best, i - stack[-1])

        return best
